File: 05-instruction-tuned-domain-llm/src/prompt_templates.py
"""Prompt templates shared by training and inference."""
from __future__ import annotations

from typing import Dict

SYSTEM_SCOPE = (
    "You are an educational ML and Data Science learning assistant. "
    "Answer clearly, stay within ML/DS, state uncertainty, and do not provide "
    "legal, medical, financial, immigration, or safety-critical advice."
)

CATEGORY_GUIDANCE: Dict[str, str] = {
    "Concept explanation": "Explain the concept in plain language, then give one practical example.",
    "Algorithm comparison": "Compare the methods using strengths, limitations, and when to use each.",
    "Metric explanation": "Define the metric, explain its interpretation, and mention one limitation.",
    "Example generation": "Provide a small, concrete example and explain what it demonstrates.",
    "Beginner-friendly explanation": "Use simple language and avoid unnecessary jargon.",
    "Interview-style answer": "Give a concise, structured answer suitable for an interview.",
    "Small code example": "Provide a minimal Python example and explain the key lines.",
    "Data Science workflow": "Describe the steps in order and include validation considerations.",
    "ML project guidance": "Provide practical project guidance with clear next steps.",
    "Quality analytics": "Use a non-confidential quality analytics example and explain the modeling choice.",
}


def clean_text(value: object) -> str:
    return " ".join(str(value or "").strip().split())


def build_training_prompt(instruction: str, input_text: str = "", category: str = "") -> str:
    instruction = clean_text(instruction)
    input_text = clean_text(input_text)
    category = clean_text(category)
    parts = [f"System: {SYSTEM_SCOPE}"]
    if category:
        guidance = CATEGORY_GUIDANCE.get(category, "Answer the instruction accurately and clearly.")
        parts.append(f"Category: {category}\nGuidance: {guidance}")
    parts.append(f"Instruction: {instruction}")
    if input_text:
        parts.append(f"Input: {input_text}")
    parts.append("Response:")
    return "\n\n".join(parts)


def build_inference_prompt(instruction: str, input_text: str = "", category: str = "") -> str:
    return build_training_prompt(instruction, input_text, category)

File: 05-instruction-tuned-domain-llm/web/src/prompt-templates.js
const SYSTEM_SCOPE = [
  'You are an educational ML and Data Science learning assistant.',
  'Answer only within machine learning, data science, analytics, NLP, computer vision, MLOps, and quality analytics.',
  'Be accurate, beginner-friendly, structured, and explicit about uncertainty.',
  'Do not give legal, medical, financial, immigration, or safety-critical advice.',
].join(' ');

export const PROMPT_CATEGORIES = Object.freeze([
  'Concept explanation',
  'Algorithm comparison',
  'Metric explanation',
  'Example generation',
  'Interview-style answer',
  'Small code or pseudo-code example',
  'Data Science workflow',
  'Quality analytics example',
]);

const CATEGORY_INSTRUCTIONS = Object.freeze({
  'Concept explanation': 'Explain the concept clearly, define important terms, and include one intuitive example.',
  'Algorithm comparison': 'Compare the methods using assumptions, strengths, limitations, and suitable use cases.',
  'Metric explanation': 'Define the metric, explain when it is useful, and include a practical interpretation.',
  'Example generation': 'Generate a small, concrete educational example and explain each step.',
  'Interview-style answer': 'Give a concise interview-ready answer followed by one deeper technical note.',
  'Small code or pseudo-code example': 'Provide a short, safe code or pseudo-code example and explain what it demonstrates.',
  'Data Science workflow': 'Describe the workflow in ordered stages, including validation and evaluation.',
  'Quality analytics example': 'Connect the explanation to defect, case, process, supplier, or quality trend analysis without using confidential data.',
});

export function buildPrompt({ category, instruction, context = '' }) {
  const cleanInstruction = instruction.trim();
  const cleanContext = context.trim();
  const categoryGuidance = CATEGORY_INSTRUCTIONS[category] ?? CATEGORY_INSTRUCTIONS['Concept explanation'];

  return [
    `Context: ${SYSTEM_SCOPE}`,
    `Task style: ${categoryGuidance}`,
    '',
    'Instruction:',
    cleanInstruction,
    '',
    'Input:',
    cleanContext || 'No additional input provided.',
    '',
    'Response:',
  ].join('\n');
}
